synapse/third_party_rules/access_rules.py

```diff
@@ -446,8 +449,136 @@ def _on_membership_or_invite(
             # might want to change that in the future.
             ret = self._on_membership_or_invite_restricted(event)
 
+        if event.type == "m.room.member":
+            # If this is an admin leaving, and they are the last admin in the room,
+            # raise the power levels of the room so that the room is 'frozen'.
+            #
+            # We have to freeze the room by puppeting an admin user, which we can
+            # only do for local users
+            if (
+                self._is_local_user(event.sender)
+                and event.membership == Membership.LEAVE
+            ):
+                await self._freeze_room_if_last_admin_is_leaving(event, state_events)
+
         return ret
 
+    async def _freeze_room_if_last_admin_is_leaving(
+        self, event: EventBase, state_events: StateMap[EventBase]
+    ):
+        power_level_state_event = state_events.get(
+            (EventTypes.PowerLevels, "")
+        )  # type: EventBase
+        if not power_level_state_event:
+            return
+        power_level_content = power_level_state_event.content.copy()
+
+        # Do some validation checks on the power level state event
+        if (
+            not isinstance(power_level_content, dict)
+            or "users" not in power_level_content
+            or not isinstance(power_level_content["users"], dict)
+        ):
+            # We can't use this power level event to determine whether the room should be
+            # frozen. Bail out.
+            return
+
+        user_id = event.get("sender")
+        if not user_id:
+            return
+
+        # Get every admin user defined in the room's state
+        admin_users = {
+            user
+            for user, power_level in power_level_content["users"].items()
+            if power_level >= 100
+        }
+
+        if user_id not in admin_users:
+            # This user is not an admin, ignore them
+            return
+
+        # Filter these users to only those who are actually joined or invited to the room
+        joined_members = [
+            user_id
+            for (event_type, user_id), event in state_events.items()
+            if event_type == EventTypes.Member
+            and event.membership in [Membership.JOIN, Membership.INVITE]
+        ]
+        admin_users = {user for user in admin_users if user in joined_members}
+
+        if len(admin_users) > 1:
+            # There's another admin user in, or invited to, the room
+            return
```

Comment:
This feels like it's doing a bunch of redundant work. We can bail out as soon as we find an admin member, and we don't need to maintain big lists/sets.

```suggestion
        if any(
            event_type == EventTypes.Member
            and event.membership in [Membership.JOIN, Membership.INVITE]
            and state_key in admin_users
            and state_key != user_id
        ):
            # There's another admin user in, or invited to, the room
            return
```

Reply:
It's actually very rare that I use `any` in my code, but I see the usefulness of it now, thanks!